pkg/reconciler/simpledeployment/simpledeployment.go

```diff
@@ -0,0 +1,169 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package simpledeployment
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/kubernetes"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+
+	"knative.dev/pkg/kmeta"
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/reconciler"
+	"knative.dev/sample-controller/pkg/apis/samples"
+	samplesv1alpha1 "knative.dev/sample-controller/pkg/apis/samples/v1alpha1"
+	simpledeploymentreconciler "knative.dev/sample-controller/pkg/client/injection/reconciler/samples/v1alpha1/simpledeployment"
+)
+
+// podOwnerLabelKey is the key to a label that points to the owner (creator) of the
+// pod, allowing us to easily list all pods a single SimpleDeployment created.
+const podOwnerLabelKey = samples.GroupName + "/podOwner"
+
+// Reconciler implements simpledeploymentreconciler.Interface for
+// SimpleDeployment resources.
+type Reconciler struct {
+	kubeclient kubernetes.Interface
+	podLister  corev1listers.PodLister
+}
+
+// Check that our Reconciler implements Interface
+var _ simpledeploymentreconciler.Interface = (*Reconciler)(nil)
+
+// ReconcileKind implements Interface.ReconcileKind.
+func (r *Reconciler) ReconcileKind(ctx context.Context, d *samplesv1alpha1.SimpleDeployment) reconciler.Event {
+	// This logger has all the context necessary to identify which resource is being reconciled.
+	logger := logging.FromContext(ctx)
+
+	// Get all the pods created by the current SimpleDeployment. The result is read from
+	// cache (via the lister).
+	selector := labels.SelectorFromSet(labels.Set{
+		podOwnerLabelKey: d.Name,
+	})
+	existingPods, err := r.podLister.Pods(d.Namespace).List(selector)
+	if err != nil {
+		return fmt.Errorf("failed to list existing pods: %w", err)
+	}
+	logger.Infof("Found %d pods in total", len(existingPods))
+
+	// Find out which pods have the current image and which ones are outdated.
+	currentPods, outdatedPods := partitionPods(existingPods, d.Spec.Image)
```

Comment:
I used a neat trick for this in net-contour where I put the generation of the spec used to create the http proxies in a label.  I was updating the http proxies (and the label), which you don't want, but what that lets you do here is leverage the lister's cache to list only the latest generation, and DeleteCollection with `!=` to clean up older generations.

Although now that I've written this, since spec includes your replica count, this would mean creating N+1 new pods for N -> N+1 🤣

Reply:
Thought about implementing the scale subresource?

Reply:
I've chosen the most straightforward thing here, seeing as this is supposed to show a (maybe naive) way of doing things.

I can add a comment stating that this is a naive approach on purpose though.